v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/source/reader/io/transform/AccumulatingTableReader.java

```diff
@@ -0,0 +1,125 @@
+import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.SerializableFunction;
+import org.apache.beam.sdk.transforms.SimpleFunction;
+import org.apache.beam.sdk.values.PBegin;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionList;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+
+/**
+ * Helper to {@link ReaderTransform} that expands the Pipeline Graph to flatten all the tableReaders
+ * and generate table completions. Please refer to {@link ReaderTransform} for greater details.
+ */
+@AutoValue
+abstract class AccumulatingTableReader extends PTransform<PBegin, PCollectionTuple> {
+  abstract ImmutableMap<SourceTableReference, PTransform<PBegin, PCollection<SourceRow>>>
+      tableTransforms();
+
+  abstract TupleTag<SourceRow> sourceRowTag();
+
+  abstract TupleTag<SourceTableReference> sourceTableReferenceTag();
+
+  @Override
+  public PCollectionTuple expand(PBegin input) {
+    ImmutableMap<SourceTableReference, PCollection<SourceRow>> tablePCollections =
+        this.tableTransforms().entrySet().stream()
+            .collect(
+                ImmutableMap.toImmutableMap(
+                    Entry::getKey,
+                    e -> input.apply("SourceRowReader." + e.getKey().getName(), e.getValue())));
+    PCollection<SourceRow> sourceRowPCollection =
+        PCollectionList.of(tablePCollections.values())
+            /* Flatten SourceRows read from all the tables into a single PCollection */
+            .apply("SourceRowCollector", Flatten.<SourceRow>pCollections());
+
+    PCollection<SourceTableReference> tableCompletions =
+        PCollectionList.of(
+                tablePCollections.entrySet().stream()
+                    .map(
+                        e ->
+                            /*
+                             * Reduce the Collection of SourceRows to their counts in a parallel stage.
+                             * This ensures that this parallel stage waits till the table is completed,
+                             * allowing the reader to signal completion of reading a table to the pipeline controller.
+                             * As needed, more complex reduce operation can be performed providing additional information to the pipelineController.
+                             * TODO(vardhanvthigle): Add a note on cost vs performance of the Count transform after benchmarking.
+                             */
+                            e.getValue()
+                                .apply("TableRowCounter." + e.getKey().getName(), Count.globally())
+                                .apply(
+                                    "TableCompletion." + e.getKey().getName(),
+                                    MapElements.via(new SourceTableReferenceWithCount(e.getKey()))))
+                    .collect(ImmutableList.toImmutableList()))
```

Comment:
I didn't get what this is trying to do, could you please elaborate a little bit for my understanding?

Can you also elaborate what do you mean by a signal - my understanding is that the table completion signal can be used for further processing - but how will it work - I didn't get that.

Won't the same signal also be needed for the DAG based implementation?

Reply:
`my understanding is that the table completion signal can be used for further processing ` 
That's absolutely right!



`how will it work `:
A small summary of the other code in this file
So the reader transform returns a tuple-tagged output with 2 tags:
1. Tag for `PCollection<SourceTableReference>` - which is the completions
2. Tag for `PCollection<SourceRows>` - which is the rows
The pipeline controller can do `pcs.get(tagForCompeltions)` this gives the PCollection<SourceTableReference> an element is added here only when a table compltes.

So now the pipeline controller needs to do `pcs.get(tagForCompeltions).apply(PTrasnformThatHandlesCompltions<SourceTableReference, T>)`

Now this `PTrasnformThatHandlesCompltions` can do whatever is needed - like say send a pubsub event to live pipeline to consume cdcs, write to spanner for checkpointing, just emit a log 


On sequencing reads between tables, that can be done here too, but it's much easier to do that natively in ReaderTransform as disccused in the previous comment (assuming reader gets the topologically sorted sequence of tables from pipeline controller based on spanner schema interleaving or spanner foreign keys)

Reply:
As discussed, will add a note on this in the design.